onnx/defs/parser.h

```diff
@@ -0,0 +1,382 @@
+#define CHECK_PARSER_STATUS(status) \
+  {                                 \
+    auto local_status_ = status;    \
+    if (!local_status_.IsOK())      \
+      return local_status_;         \
+  }
+
+class PrimitiveTypeNameMap {
+ public:
+  PrimitiveTypeNameMap() {
+    map_["float"] = 1;
+    map_["uint8"] = 2;
+    map_["int8"] = 3;
+    map_["uint16"] = 4;
+    map_["int16"] = 5;
+    map_["int32"] = 6;
+    map_["int64"] = 7;
+    map_["string"] = 8;
+    map_["bool"] = 9;
+    map_["float16"] = 10;
+    map_["double"] = 11;
+    map_["uint32"] = 12;
+    map_["uint64"] = 13;
+    map_["complex64"] = 14;
+    map_["complex128"] = 15;
+    map_["bfloat16"] = 16;
+  }
+
+  static const std::unordered_map<std::string, int32_t>& Instance() {
+    static PrimitiveTypeNameMap instance;
+    return instance.map_;
+  }
+
+  static int32_t Lookup(const std::string& dtype) {
+    auto it = Instance().find(dtype);
+    if (it != Instance().end())
+      return it->second;
+    return 0;
+  }
+
+  static bool IsTypeName(const std::string& dtype) {
+    return Lookup(dtype) != 0;
+  }
+
+  static const std::string& ToString(int32_t dtype) {
+    static std::string undefined("undefined");
+    for (const auto& pair : Instance()) {
+      if (pair.second == dtype)
+        return pair.first;
+    }
+    return undefined;
+  }
+
+ private:
+  std::unordered_map<std::string, int32_t> map_;
+};
+
+class KeyWordMap {
+ public:
+  enum class KeyWord {
+    NONE,
+    IR_VERSION,
+    OPSET_IMPORT,
+    PRODUCER_NAME,
+    PRODUCER_VERSION,
+    DOMAIN_KW,
+    MODEL_VERSION,
+    DOC_STRING,
+    METADATA_PROPS
+  };
+
+  KeyWordMap() {
+    map_["ir_version"] = KeyWord::IR_VERSION;
+    map_["opset_import"] = KeyWord::OPSET_IMPORT;
+    map_["producer_name"] = KeyWord::PRODUCER_NAME;
+    map_["producer_version"] = KeyWord::PRODUCER_VERSION;
+    map_["domain"] = KeyWord::DOMAIN_KW;
+    map_["model_version"] = KeyWord::MODEL_VERSION;
+    map_["doc_string"] = KeyWord::DOC_STRING;
+    map_["metadata_props"] = KeyWord::METADATA_PROPS;
+  }
+
+  static const std::unordered_map<std::string, KeyWord>& Instance() {
+    static KeyWordMap instance;
+    return instance.map_;
+  }
+
+  static KeyWord Lookup(const std::string& id) {
+    auto it = Instance().find(id);
+    if (it != Instance().end())
+      return it->second;
+    return KeyWord::NONE;
+  }
+
+ private:
+  std::unordered_map<std::string, KeyWord> map_;
+};
+
+class ParserBase {
+ public:
+  ParserBase(const std::string& str)
+      : start_(str.data()), next_(str.data()), end_(str.data() + str.length()), saved_pos_(next_) {}
+
+  ParserBase(const char* cstr) : start_(cstr), next_(cstr), end_(cstr + strlen(cstr)), saved_pos_(next_) {}
+
+  void SavePos() {
+    saved_pos_ = next_;
+  }
+
+  void RestorePos() {
+    next_ = saved_pos_;
+  }
+
+  std::string GetCurrentPos() {
+    uint32_t line = 1, col = 1;
+    for (const char* p = start_; p < next_; p++) {
+      if (*p == '\n') {
+        line++;
```

Comment:
In situations like this it is preferred to use `++line` rather than `line++`. The same applies to the for-loop statements increasing a counter/pointer.

Reply:
Done. But it is a pity if a compiler cannot optimize this in this day and age :)